### ISSUE TITLE
no-undef fails with node.js global object

### ISSUE THREAD

**Issue by LJNeon** [eslint/eslint#11333]
**Tell us about your environment**

* **ESLint Version:** 5.12.1
* **Node Version:** 10.15.1
* **npm Version:** 6.4.1

**What parser (default, Babel-ESLint, etc.) are you using?**
default
**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
{
  "env": {
    "node": true
  },
  "extends": "eslint:recommended",
  "parserOptions": {
    "ecmaVersion": 2018
  }
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
global.my_value = 5;

if (my_value < 3) {
  // ...
}
```

```bash
eslint . --ext .js
```

**What did you expect to happen?**
No error would occur as the nodejs environment is set.

**What actually happened? Please include the actual, raw output from ESLint.**
```
/path/to/my/index.js
   3:5  error  'my_value' is not defined      no-undef
```

**Are you willing to submit a pull request to fix this bug?**
No, I lack the experience with eslint internal code to reliably do so.

**Comment by g-plane** [comment]
Your variable `my_value` is really not defined. This isn't a bug.

**Comment by ljharb** [comment]
@g-plane that's not correct; assigning properties to the global makes them also available as global variables.

**Comment by g-plane** [comment]
Yes it's OK at runtime but invalid for static checking.

**Comment by g-plane** [comment]
The solution is adding the following line at the header of the file:

```javascript
/* global my_value */
```

**Comment by ljharb** [comment]
I believe the issue is that eslint is capable of statically determining that this global is created, and inferring that global setting.

**Comment by not-an-aardvark** [comment]
Personally I don't think this change would be worth the complexity, because:

* ESLint currently isn't aware that `global` represents the global object in Node. The `node` env just enables some specific globals such as `module`, `global`, and `exports` (and also some unrelated parser options), but there isn't any concept of "assignments to this value will create a property on the global object".
* This won't create a global if someone has reassigned `global.global` to something other than the global object.
* Determining whether a particular global is present at any given point is undecidable, since a property could also be conditionally deleted from the global object (or deleted in another file). On the other hand, determining whether a non-global variable is in scope is statically decidable for strict mode code.

**Comment by ljharb** [comment]
Reasonable! Just wanted to clarify that it is, in fact, creating a statically determinable global variable.

**Comment by bfred-it** [comment]
Could this picked up again once [proposal-global](https://github.com/tc39/proposal-global) makes it into ES?

https://github.com/tc39/proposal-global

**Comment by ljharb** [comment]
@bfred-it i'm not clear on how my proposal would impact this issue. Creating a property on an environment-specific global object is easily handled by an `env` setting; doing so on `globalThis` would land in some future ecmaVersion - but this rule would either handle all such cases, or none, and I'm not sure why it would make sense to do anything in the middle?

**Comment by bfred-it** [comment]
> * This won't create a global if someone has reassigned `global.global` to something other than the global object.

> 
> * Determining whether a particular global is present at any given point is undecidable, since a property could also be conditionally deleted from the global object (or deleted in another file). On the other hand, determining whether a non-global variable is in scope is statically decidable for strict mode code.

**Comment by ljharb** [comment]
@bfred-it "proper" globals can basically all be reassigned; `setTimeout` can certainly be reassigned as well (see `Object.getOwnPropertyDescriptor(window, 'setTimeout')` everywhere) - as can `globalThis`.

**Comment by bfred-it** [comment]
> @bfred-it "proper" globals can basically all be reassigned; `setTimeout` can certainly be reassigned as well (see `Object.getOwnPropertyDescriptor(window, 'setTimeout')` everywhere) - as can `globalThis`.

Right, that only adds more weight to what I'm saying. Those two points apply to all globals, yet eslint is fine with `setTimeout` being a global (that can be overridden) but not other locally-defined globals.

**Comment by bfred-it** [comment]
TypeScript now also detects variables added on `globalThis`, which shows it can be done, if desired.

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#type-checking-for-globalthis
